### PR TITLE
Fast FixedPrecisonTensor Reciprocal Method (150x faster)

### PR DESCRIPTION
**Pull request by marload**
## Description

issue: https://github.com/OpenMined/PySyft/issues/3993

We can use reciprocal 150 times faster on average. Implementing division of AST with reciprocal can make existing very slow division much faster. In addition, many slow operations can be made very quickly.

This implementation referred to the implementation of CrypTen.

https://github.com/facebookresearch/CrypTen/blob/3c4e53aa4910117e4e136691cb09ea7a7e001029/crypten/mpc/mpc.py#L1029

```python
x = torch.tensor([1., 2., 3.]).fix_prec().share(bob, alice, crypto_provider=theo)

# Existing method (using division)
x.reciprocal().get().float_prec() # 27 seconds
# tensor([1.0000, 0.5000, 0.3330])

# New Method (using NR)
x.reciprocal(method="nr").get().float_prec() # 0.18 seconds
# tensor([1.0000, 0.5000, 0.3330])

# New Method (using LOG) / Approximation
x.reciprocal(method="log").get().float_prec() # 0.31 seconds
#  tensor([1.0000, 0.4650, 0.3530])
```

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: test/torch/tensors/test_precision.py [hunk 1]
```diff
@@ -103,6 +103,22 @@ def test_methods_for_linear_module(method, parameter):
     assert (result == fp_result.float_precision()).all()
 
 
+def test_reciprocal(workers):
+    bob, alice, james = (workers["bob"], workers["alice"], workers["james"])
+
+    x = torch.tensor([1.0, 2.0, 3.0]).fix_prec()
```
Comment (gmuraru) [inline test/torch/tensors/test_precision.py:109]: MAYBE: Use 
```
tensor = torch.tensor([1.0, 2.0, 3.0])
x = tensor.fix_prec()

result = x.reciprocal().float_prec()
assert (result == tensor.reciprocal()).all()
```

Comment (marload) [reply test/torch/tensors/test_precision.py:109]: Because the torch tensor uses the Floating Point, the original method of using division also has an error. That's why you can't use this method. 😄 

```python
tensor = torch.tensor([1.0, 2.0, 3.0])
x = tensor.fix_prec()

result = x.reciprocal().float_prec()
assert (result == tensor.reciprocal()).all() # -> tensor([True, True, False])
# tensor([1.0000, 0.5000, 0.3333]) == tensor([1.0000, 0.5000, 0.3333]) == 
```

Comment (gmuraru) [reply test/torch/tensors/test_precision.py:109]: Ahh...don't use equal, use: [```torch.isclose```](https://pytorch.org/docs/stable/generated/torch.isclose.html)

Comment (marload) [reply test/torch/tensors/test_precision.py]: I applied this change request and added a docstring 😄

**Comment by gmuraru** [comment]
Re-run the tests. It might be a problem regarding the ```seed``` - we might need to increase the tolerance (this issue was seen before :\)

**Comment by Syzygianinfern0** [comment]
> Re-run the tests. It might be a problem regarding the `seed` - we might need to increase the tolerance (this issue was seen before :)

It ran 3 times and failed all of them :rofl:

**Comment by marload** [comment]
@Syzygianinfern0 @gmuraru 

The error appears to be caused by a small error in the NR. We solved this problem by setting the default value of reciprocal to division. :smile

**Comment by Syzygianinfern0** [comment]
Is the NR scheme buggy? Why did it fail for that?

> @Syzygianinfern0 @gmuraru
> 
> The error appears to be caused by a small error in the NR. We solved this problem by setting the default value of reciprocal to division.

But how does an error tolerance of about `1e-2` (as seen from the test cases you've written) propagate to something like `20 < 1` ?

**Comment by marload** [comment]
NR uses the Approx Exp, so there is a little more error than Division. I think this problem is caused by the error because `tanh` uses `_sigmoid_exp`.

**Comment by gmuraru** [comment]
If we change such that ```sigmoid``` uses ```chebyshev``` by default, does it work?

**Comment by marload** [comment]
Yes, that's right. No errors occur in the `chebyshev`.

**Comment by marload** [comment]
If `chebyshev` is the default, no error occurs.
I just checked. 😄

**Comment by gmuraru** [comment]
I think we can go with ```chebyshev``` as the default one

**Comment by Syzygianinfern0** [comment]
@marload 
A couple of things to consider for future PRs :smile: 
- Squashing basic commits to hold a good commit structure before a push
- Working on a branch different from the master on your fork (makes it easier for you)

Thank you again for your contributions!
